internal/kafka/circuitbreaker.go

```diff
@@ -0,0 +1,184 @@
+// Copyright 2018, 2021 Kaleido
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kafka
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"github.com/hyperledger/firefly-ethconnect/internal/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	DefaultUpperBound     int64   = 80 * (1024 * 1024)
+	DefaultResetThreshold float64 = 0.8
+	DefaultLogFrequency           = 30 * time.Second
+)
+
+// CircuitBreakerConf defines the YAML config structure for the circuit breaker
+type CircuitBreakerConf struct {
+	Enabled         bool    `json:"enabled,omitempty"`
+	UpperBound      int64   `json:"upperBound,omitempty"`
+	ResetThreshold  float64 `json:"resetFraction,omitempty"`
+	LogFrequencySec int     `json:"logFrequencySec,omitempty"`
+
+	// calculated values
+	logFrequency    time.Duration
+	resetBufferSize int64
+}
+
+type CircuitBreaker interface {
+	Update(topic string, partition int32, hwm, offset, size int64)
+	Check(topic string) error
+}
+
+// cbPartitionState is a per-topic state structure for managing trips
+type cbPartitionState struct {
+	sizeEstimate int64
+	offset       int64
+	hwm          int64
+	gap          int64
+	bufSize      int64
+	msgCount     int64
+	msgBytes     int64
+	tripped      bool
+	lastTripTime time.Time
+	lastLogged   time.Time
+}
+
+// circuitBreaker checks that the
+type circuitBreaker struct {
+	conf  *CircuitBreakerConf
+	mux   sync.Mutex
+	state map[string]map[int32]*cbPartitionState
```

Comment:
Can we add a comment that describes the mapping `{topic} -> {partition} -> {partitionState}` is what it appears to be, based on reading the code further down.